server/athenian/api/controllers/invitation_controller.py

```diff
@@ -0,0 +1,160 @@
+        salt = existing[Invitation.salt.key]
+    else:
+        # create a new invitation
+        salt = randint(0, (1 << 16) - 1)  # 0:65535 - 2 bytes
+        inv = Invitation(salt=salt, account_id=id, created_by=request.user.id).create_defaults()
+        invitation_id = await sdb.execute(insert(Invitation).values(inv.explode()))
+    slug = encode_slug(invitation_id, salt)
+    model = InvitationLink(url=prefix + slug)
+    return response(model)
+
+
+def encode_slug(iid: int, salt: int) -> str:
+    """Encode an invitation ID and some extra data to 8 chars."""
+    part1 = struct.pack("!H", salt)  # 2 bytes
+    part2 = struct.pack("!I", iid)[1:]  # 3 bytes
+    binseq = (part1 + part2).hex()  # 5 bytes, 10 hex chars
+    e = pyffx.String(ikey.encode(), alphabet="0123456789abcdef", length=len(binseq))
+    encseq = e.encrypt(binseq)  # encrypted 5 bytes, 10 hex chars
+    finseq = base64.b32encode(bytes.fromhex(encseq)).lower().decode()  # 8 base32 chars
+    finseq = finseq.replace("o", "8").replace("l", "9")
+    return finseq
+
+
+def decode_slug(slug: str) -> (int, int):
+    """Decode an invitation ID and some extra data from 8 chars."""
+    assert len(slug) == 8
+    assert isinstance(slug, str)
+    b32 = slug.replace("8", "o").replace("9", "l").upper().encode()
+    x = base64.b32decode(b32).hex()
+    e = pyffx.String(ikey.encode(), alphabet="0123456789abcdef", length=len(x))
+    x = bytes.fromhex(e.decrypt(x))
+    salt = struct.unpack("!H", x[:2])[0]
+    iid = struct.unpack("!I", b"\x00" + x[2:])[0]
+    return iid, salt
+
+
+async def accept_invitation(request: AthenianWebRequest, body: dict) -> web.Response:
+    """Accept the membership invitation."""
+    def bad_req():
+        return ResponseError(BadRequestError(detail="Invalid invitation URL")).response
+
+    sdb = request.sdb
+    url = InvitationLink.from_dict(body).url
+    if not url.startswith(prefix):
+        return bad_req()
+    x = url[len(prefix):].strip("/")
+    if len(x) != 8:
+        return bad_req()
+    try:
+        iid, salt = decode_slug(x)
+    except binascii.Error:
+        return bad_req()
+    async with sdb.connection() as conn:
+        inv = await conn.fetch_one(
+            select([Invitation.account_id, Invitation.accepted, Invitation.is_active])
+            .where(and_(Invitation.id == iid, Invitation.salt == salt)))
+        if inv is None:
+            return ResponseError(NotFoundError(
+                detail="Invitation was not found.")).response
+        if not inv[Invitation.is_active.key]:
+            return ResponseError(ForbiddenError(
+                detail="This invitation is disabled.")).response
+        acc = inv[Invitation.account_id.key]
+        is_admin = acc == admin_backdoor
+        if is_admin:
+            # create a new account for the admin user
+            acc = await conn.execute(insert(Account).values(Account().create_defaults().explode()))
+            if acc == admin_backdoor:
```

Comment:
I think that this should be:

```suggestion
            if acc[Invitation.account_id.key] == admin_backdoor:
```

Reply:
This one no, the insertion result is an integer primary key. IDK how all the underlying magic works but that's a tested fact...